### ISSUE TITLE
task suspend and resume

### ISSUE THREAD

**Issue by yonik** [taskflow/taskflow#291]
### General Problem
Although taskflow is only a framework for compute, there still needs to be mechanisms to allow better coexistence/integration with IO driven frameworks.  A task may not have all of the resources needed to complete its computation at the time it is scheduled to run, and one does not always know ahead of time if this will be the case. 

Related: TBB now has [resumable tasks](https://spec.oneapi.com/versions/latest/elements/oneTBB/source/task_scheduler/scheduling_controls/resumable_tasks.html)
and even more relevant [async_node](https://spec.oneapi.io/versions/latest/elements/oneTBB/source/flow_graph/async_node_cls.html)
Related: https://github.com/taskflow/taskflow/issues/126

### Possible Implementations
I'm *not* suggesting implementing something like the TBB feature (suspend/resume at an arbitrary point in a task), since that would seem to require platform specific non-portable code to save and restore stack and registers.

At a high level, all that is needed to allow the user to integrate other arbitrary frameworks is a way to change the state of a task to a "suspended" state (i.e. don't try to schedule this task), and a way to later un-suspend (i.e. tell taskflow that the task can now be scheduled again.)

To avoid having to implement suspend-anywhere,  a simple implementation would be to have a call that informed the scheduler that the current task should be "suspended" after it returns.  A later call outside of taskflows domain would inform taskflow that the task can now be scheduled (restarted from the beginning.)

**Comment by tsung-wei-huang** [comment]
Probably this should be done at the taskflow level not task  - in many task parallel frameworks a task is the basic execution unit. Without importing coroutine-like concurrency, it's not possible to pause a task.

**Comment by yonik** [comment]
> it's not possible to pause a task

I assumed you already had something like that to integrate CPU and GPU tasks.  If a CPU task depends on a GPU task, the CPU task can't be scheduled until the GPU task completes.  Seems like all we need is something external to taskflow to be able to say "OK, now this task can be scheduled".

**Comment by yonik** [comment]
I figured out a work-around using a tf::Semaphore initialized to 0. Then an external system can signal/release the semaphore via running a dummy task for that purpose.

```
    // example code inside a task already
    auto [A,B] = subflow.emplace(
            [](){std::cout << "\tTask A" << std::endl;},
            [](){std::cout << "\tTask B" << std::endl;}
    );

    // Now we decide that A needs a resource we don't yet have (and getting it could block)
    A.acquire(job->sem);  // sem starts out at 0, so this will prevent A from running
    std::thread async([this,job](){
      std::this_thread::sleep_for(std::chrono::milliseconds(5000));
      // We can't use the semaphore directly, but we can run a task that releases it
      tf::Taskflow taskflow;
      auto dummy = taskflow.emplace([](){}); // TODO: how to set highest priority for this task?
      dummy.release(job->sem); // dummy's only task is to signal the semaphore
      executor->run(std::move(taskflow));
      std::cout << "executor running wakeup task" << std::endl;
    });
    async.detach();
```

Any suggestions on how to improve this signaling mechanism? Would it be possible to allow explicitly releasing a semaphore outside of a worker?  One thing I noticed is that this does cause the CPU to spin if there are no other tasks to execute (on taskflow 3.3 at least)... is that expected?

**Comment by tsung-wei-huang** [comment]
closed for now as there are no more activities.
